File: frontend/src/services/retentionService.js
/**
 * retentionService.js – API client for data retention policy management
 */

import api from '../lib/api'

const RETENTION_API_BASE = '/api/forms'

/**
 * Get retention policy for a form
 * @param {string} formId - Form UUID
 * @returns {Promise<Object>} Retention policy data
 */
export async function getRetentionPolicy(formId) {
  try {
    const response = await api.get(`${RETENTION_API_BASE}/${formId}/retention-policy`)
    return response.data
  } catch (error) {
    console.error('Failed to fetch retention policy:', error)
    throw error
  }
}

/**
 * Update retention policy for a form
 * @param {string} formId - Form UUID
 * @param {Object} policy - Policy data (enabled, retention_days, action)
 * @returns {Promise<Object>} Updated retention policy
 */
export async function updateRetentionPolicy(formId, policy) {
  try {
    const response = await api.patch(`${RETENTION_API_BASE}/${formId}/retention-policy`, policy)
    return response.data
  } catch (error) {
    console.error('Failed to update retention policy:', error)
    throw error
  }
}

/**
 * Manually trigger retention archival job (admin/testing only)
 * @returns {Promise<Object>} Archival result (forms_processed, responses_archived, execution_time_ms)
 */
export async function triggerRetentionArchival() {
  try {
    const response = await api.post('/admin/retention/run')
    return response.data
  } catch (error) {
    console.error('Failed to trigger retention archival:', error)
    throw error
  }
}

File: backend/app/services/retention_service.py
"""Service for managing data retention policies and automatic response archival."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.form_platform import Form, Response
from app.models.retention_policies import RetentionPolicy
from app.services.audit_log_service import AuditLogService


class RetentionService:
    """Service for retention policy management and archival operations."""

    @staticmethod
    def get_or_create_retention_policy(
        db: Session,
        form_id: UUID,
    ) -> RetentionPolicy:
        """Get existing retention policy for a form, or create a default one."""
        policy = db.scalars(
            select(RetentionPolicy).where(RetentionPolicy.form_id == form_id)
        ).first()

        if policy is not None:
            return policy

        # Create default policy (disabled, 90 days)
        policy = RetentionPolicy(
            form_id=form_id,
            enabled=False,
            retention_days=90,
            action="archive",
        )
        db.add(policy)
        db.commit()
        db.refresh(policy)
        return policy

    @staticmethod
    def update_retention_policy(
        db: Session,
        form_id: UUID,
        enabled: bool | None = None,
        retention_days: int | None = None,
        action: str | None = None,
    ) -> RetentionPolicy:
        """Update retention policy for a form."""
        policy = RetentionService.get_or_create_retention_policy(db, form_id)

        if enabled is not None:
            policy.enabled = enabled
        if retention_days is not None:
            policy.retention_days = retention_days
        if action is not None:
            policy.action = action

        db.add(policy)
        db.commit()
        db.refresh(policy)
        return policy

    @staticmethod
    def archive_expired_responses(db: Session) -> dict[str, int]:
        """
        Archive all responses that have exceeded their retention period.

        Returns:
            {
                "forms_processed": number of forms with enabled policies checked,
                "responses_archived": total responses archived,
            }
        """
        forms_processed = 0
        total_archived = 0

        # Get all forms with enabled retention policies
        forms = db.scalars(
            select(Form)
            .join(RetentionPolicy, Form.id == RetentionPolicy.form_id)
            .where(RetentionPolicy.enabled == True)
        ).all()

        for form in forms:
            policy = form.retention_policy
            if not policy or not policy.enabled:
                continue

            forms_processed += 1

            # Calculate cutoff date: submitted_at < now - retention_days
            cutoff_date = datetime.now(timezone.utc) - timedelta(days=policy.retention_days)

            # Find responses to archive
            responses_to_archive = db.scalars(
                select(Response)
                .where(Response.form_id == form.id)
                .where(Response.status == "submitted")
                .where(Response.submitted_at < cutoff_date)
            ).all()

            if not responses_to_archive:
                continue

            # Archive responses by setting status
            for response in responses_to_archive:
                response.status = "archived"

            db.commit()
            archived_count = len(responses_to_archive)
            total_archived += archived_count

            # Create audit log entry
            AuditLogService.create_audit_log(
                db=db,
                user_id=None,  # System action
                action="RETENTION_ARCHIVE_RESPONSES",
                form_id=form.id,
                resource_type="response",
                details={
                    "responses_archived": archived_count,
                    "retention_days": policy.retention_days,
                    "cutoff_date": cutoff_date.isoformat(),
                },
            )

        return {
            "forms_processed": forms_processed,
            "responses_archived": total_archived,
        }
